ksql-api/src/main/java/io/confluent/ksql/api/server/BufferedPublisher.java

```diff
@@ -0,0 +1,284 @@
+  private boolean cancelled;
+  private Runnable drainHandler;
+  private boolean complete;
+  private boolean completing;
+
+  /**
+   * Construct a BufferedPublisher
+   *
+   * @param ctx The Vert.x context to use for the publisher - the publisher code must always be
+   *            executed on this context. This ensures the code is never executed concurrently by
+   *            more than one thread.
+   */
+  public BufferedPublisher(final Context ctx) {
+    this(ctx, Collections.emptySet(), DEFAULT_BUFFER_MAX_SIZE);
+  }
+
+  /**
+   * Construct a BufferedPublisher
+   *
+   * @param ctx           The Vert.x context to use for the publisher
+   * @param initialBuffer A collection of elements to initialise the buffer with
+   */
+  public BufferedPublisher(final Context ctx, final Collection<T> initialBuffer) {
+    this(ctx, initialBuffer, DEFAULT_BUFFER_MAX_SIZE);
+  }
+
+  /**
+   * Construct a BufferedPublisher
+   *
+   * @param ctx           The Vert.x context to use for the publisher
+   * @param bufferMaxSize Indicative max number of elements to store in the buffer. Note that this
+   *                      is not enforced, but it used to determine what to return from the accept
+   *                      method so the caller can stop sending more and set a drainHandler to be
+   *                      notified when the buffer is cleared
+   */
+  public BufferedPublisher(final Context ctx, final int bufferMaxSize) {
+    this(ctx, Collections.emptySet(), bufferMaxSize);
+  }
+
+  /**
+   * Construct a BufferedPublisher
+   *
+   * @param ctx           The Vert.x context to use for the publisher
+   * @param initialBuffer A collection of elements to initialise the buffer with
+   * @param bufferMaxSize Indicative max number of elements to buffer
+   */
+  public BufferedPublisher(final Context ctx, final Collection<T> initialBuffer,
+      final int bufferMaxSize) {
+    this.ctx = ctx;
+    this.buffer.addAll(initialBuffer);
+    this.bufferMaxSize = bufferMaxSize;
+  }
+
+  /**
+   * Provide an element to the publisher. The publisher will attempt to deliver it to it's
+   * subscriber (if any). The publisher will buffer it internally if it can't deliver it
+   * immediately.
+   *
+   * @param t The element
+   * @return true if the internal buffer is 'full'. I.e. if number of elements is >= bufferMaxSize.
+   */
+  public boolean accept(final T t) {
+    checkContext();
+    if (completing) {
+      throw new IllegalStateException("Cannot call accept after complete is called");
+    }
+    if (demand == 0 || cancelled) {
+      buffer.add(t);
+    } else {
+      doOnNext(t);
+    }
+    return buffer.size() >= bufferMaxSize;
+  }
+
+  /**
+   * If you set a drain handler. It will be called if, after delivery is attempted there are zero
+   * elements buffered internally and there is demand from the subscriber for more elements.
+   *
+   * @param handler The handler
+   */
+  public void drainHandler(final Runnable handler) {
```

Comment:
we should explicitly comment that this is a "one time use" function - after called once, it won't be called without another call to `drainHandler`

Reply:
Will add clarifying comment.